### PR TITLE
Refactor deployment healthcheck

### PR DESCRIPTION
**Pull request by tejal29**
In this Pr

- Refactor health check to add a Resource interface which will implement the details if a resource is healthy or not.

- Print health check summary after every resource check is returned indicating, if the current resource is ready or in error. How many more are remaining.
```
deployment/leeroy-app is ready. [1/2 deployment(s) still pending]
```
In case the deployment is in error, the message looks like:
```
deployment/leeroy-app failed [1/2 deployment(s) still pending]. Error: resource deployment/leeroy-app could not be fetched within 10s: context deadline exceeded.
deployment/leeroy-app is pending due to resource deployment/leeroy-app could not be fetched within 10s: context deadline exceeded

```

- After every 150 milliseconds, print the state of the resource, to give users more insights on what the current status of a deployment is. e.g in case of deployments with a large number of `replicas`, 
the user will see the current stautus of the deployment
```
deployment.apps/leeroy-web configured
Waiting for deployments to stabilize
deployment/leeroy-app is pending due to Waiting for rollout to finish: 5 out of 10 new replicas have been updated...
deployment/leeroy-app is pending due to Waiting for rollout to finish: 6 out of 10 new replicas have been updated...
```

- And then finally special error handling for errors when we can't fetch deployment due to connectivity issues. In such case, we should wait until the `statusCheckDeadlineSeconds` is not reached.
```
tejaldesai@@microservices (health_check_refactor)$ skaffold dev --default-repo=gcr.io/tejal-test --status-check
Generating tags...
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web -> gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web:v0.36.0-104-gdeba6031
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app -> gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app:v0.36.0-104-gdeba6031-dirty
Tags generated in 23.115371ms
Checking cache...
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web: Found
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app: Found
Cache check complete in 889.162173ms
Starting deploy...
kubectl client version: 1.11+
kubectl version 1.12.0 or greater is recommended for use with Skaffold
deployment.apps/leeroy-web created
service/leeroy-app configured
deployment.apps/leeroy-app created
Deploy complete in 1.265634429s
Waiting for deployments to stabilize
deployment/leeroy-web is ready. [1/2 deployment(s) still pending]
deployment/leeroy-app is pending due to Waiting for rollout to finish: 0 of 3 updated replicas are available...
deployment/leeroy-app is pending due to Waiting for rollout to finish: 0 of 3 updated replicas are available...
deployment/leeroy-app is pending due to Running [kubectl --context gke_tejal-test_us-central1-a_dump rollout status deployment leeroy-app --namespace default --watch=false]: stdout , stderr: Unable to connect to the server: dial tcp 35.238.80.213:443: connect: network is unreachable
, err: exit status 1: exit status 1
deployment/leeroy-app is ready. 
Watching for changes...

```

Sample full output
```
tejaldesai@@microservices (health_check_refactor)$ skaffold dev --default-repo=gcr.io/tejal-test --status-check
Generating tags...
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web -> gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web:v0.36.0-104-gdeba6031
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app -> gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app:v0.36.0-104-gdeba6031-dirty
Tags generated in 29.447689ms
Checking cache...
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web: Found
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app: Found
Cache check complete in 739.122386ms
Starting deploy...
kubectl client version: 1.11+
kubectl version 1.12.0 or greater is recommended for use with Skaffold
deployment.apps/leeroy-web created
service/leeroy-app configured
deployment.apps/leeroy-app created
Deploy complete in 1.039965467s
Waiting for deployments to stabilize
deployment/leeroy-app is pending due to Waiting for rollout to finish: 0 of 3 updated replicas are available...
deployment/leeroy-web is pending due to Waiting for rollout to finish: 0 of 1 updated replicas are available...
deployment/leeroy-app is ready. [1/2 deployment(s) still pending]
deployment/leeroy-web is ready. 
Watching for changes...
[leeroy-app-54788f4dc4-dsdlx leeroy-app] 2019/08/28 23:43:16 leeroy app server ready
[leeroy-app-54788f4dc4-tvhqt leeroy-app] 2019/08/28 23:43:16 leeroy app server ready
[leeroy-app-54788f4dc4-g87km leeroy-app] 2019/08/28 23:43:16 leeroy app server ready
[leeroy-web-5b986bffc4-xj77k leeroy-web] 2019/08/28 23:43:16 leeroy web server ready
^C
```

### CODE REVIEW

**Comment by tejal29** [comment]
@dgageot Thanks! The pod check is upcoming in another PR, but i will show you how it look. 

Regarding the "UI changes", let me fix it.

Thanks
Tejal

**Comment by tejal29** [comment]
@dgageot This is how it looks on my branch
```
tejaldesai@@microservices (health_check_refactor)$ skaffold dev --default-repo=gcr.io/tejal-test --status-check
Generating tags...
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web -> gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web:v0.36.0-105-ga8432921
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app -> gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app:v0.36.0-105-ga8432921
Tags generated in 23.651552ms
Checking cache...
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-web: Found
 - gcr.io/tejal-test/gcr.io/k8s-skaffold/leeroy-app: Found
Cache check complete in 749.659109ms
Starting deploy...
kubectl client version: 1.11+
kubectl version 1.12.0 or greater is recommended for use with Skaffold
deployment.apps/leeroy-web created
service/leeroy-app created
deployment.apps/leeroy-app created
Deploy complete in 1.019251518s
Waiting for deployments to stabilize
 - deployment/leeroy-app is pending due to Waiting for rollout to finish: 0 of 1 updated replicas are available...
 - deployment/leeroy-web is pending due to Waiting for rollout to finish: 0 of 1 updated replicas are available...
 - deployment/leeroy-web is ready. [1/2 deployment(s) still pending]
 - deployment/leeroy-app is ready. 
Deployments stabilized in 2.178304751s
Watching for changes...

```

File: pkg/skaffold/deploy/resources/resource.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+)
+
+type ResourceObj struct {
+	name      string
+	namespace string
+	rType     string
+	status    Status
+}
+
+func (r *ResourceObj) String() string {
```
Comment (balopat) [inline pkg/skaffold/deploy/resources/resource.go]: I think namespace would be interesting here too especially in multinamespace deployments

File: pkg/skaffold/deploy/resources/deployment_test.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		expectedReason string
+		complete       bool
+	}{
+		{
+			description: "rollout status success",
+			command: testutil.NewFakeCmd(t).
+				WithRunOut(rolloutCmd, "deployment dep successfully rolled out"),
+			expectedReason: "deployment dep successfully rolled out",
+			complete:       true,
+		}, {
```
Comment (balopat) [inline pkg/skaffold/deploy/resources/deployment_test.go]: nit new line - and all the ones below too 
```suggestion
		}, 
		{
```

File: pkg/skaffold/deploy/resources/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+func TestUpdateTimestamp(t *testing.T) {
+	dep := NewDeployment("test", "test-ns", time.Millisecond)
+
+	// Check updated bool is false
+	testutil.CheckDeepEqual(t, false, dep.status.updated)
+
+	// Update the status
+	dep.UpdateStatus("success", "success", nil)
+	// Check the updated bool is true
```
Comment (balopat) [inline pkg/skaffold/deploy/resources/resource_test.go]:
```suggestion
	
	// Check the updated bool is true
```

File: pkg/skaffold/deploy/resources/resource.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+func (r *ResourceObj) Type() string {
+	return r.rType
+}
+
+func (r *ResourceObj) UpdateStatus(msg string, reason string, err error) {
+	newStatus := Status{details: msg, reason: reason, err: err}
+	if !r.status.Equals(&newStatus) {
+		r.status.err = err
+		r.status.details = strings.TrimSuffix(msg, "\n")
```
Comment (balopat) [inline pkg/skaffold/deploy/resources/resource.go]: we are using `TrimSuffix(string, "\n")` so many places maybe we can just make it into a `trim(string)` function for brevity?

File: pkg/skaffold/deploy/resources/deployment_test.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	tests := []struct {
+		description    string
+		command        util.Command
+		expectedErr    string
+		expectedReason string
+		complete       bool
+	}{
+		{
+			description: "rollout status success",
+			command: testutil.NewFakeCmd(t).
```
Comment (dgageot) [inline pkg/skaffold/deploy/resources/deployment_test.go]: This code needs to be rebased

File: pkg/skaffold/deploy/resources/resource_test.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	}{
+		{
+			description: "updating an error status",
+			message:     "cannot pull image",
+			err:         fmt.Errorf("cannot pull image"),
+			expected:    " - deployment/test is pending due to cannot pull image\n",
+		}, {
+			description: "updating a non error status",
+			message:     "is waiting for container",
+			expected:    " - deployment/test is pending due to is waiting for container\n",
```
Comment (dgageot) [inline pkg/skaffold/deploy/resources/resource_test.go]: I'd love to see ` - deployment/test is waiting for container\n`

File: pkg/skaffold/deploy/resources/resource_test.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		description string
+		message     string
+		err         error
+		expected    string
+	}{
+		{
+			description: "updating an error status",
+			message:     "cannot pull image",
+			err:         fmt.Errorf("cannot pull image"),
+			expected:    " - deployment/test is pending due to cannot pull image\n",
```
Comment (dgageot) [inline pkg/skaffold/deploy/resources/resource_test.go]: I'd love to see ` - deployment/test cannot pull image\n`

**Comment by tejal29** [comment]
@dgageot Atcually i can. 
i am breaking it up in 3 parts. 

1. Just adding final printSummary https://github.com/GoogleContainerTools/skaffold/pull/2811
2. Reporting Summary after every 100 ms https://github.com/GoogleContainerTools/skaffold/pull/2813
3. Adding Resources and refactoring.

Once those 2 merged in, i will update this PR. This Pr will be relatively shorter than.

**Pull request closed without merging**